theano/compile/builders.py

```diff
@@ -121,14 +144,110 @@ def make_node(self, *inputs):
                          [type() for type in self.output_types])
 
     def make_thunk(self, node, storage_map, compute_map, no_recycling):
-        ret = super(OpFromGraph, self).make_thunk(node, storage_map,
-                                                  compute_map, no_recycling)
-        if not hasattr(self, "fn"):
-            self.fn = orig_function(self.new_inputs,
-                                    self.new_outputs,
-                                    **self.kwargs)
+        ret = super(OpFromGraph, self).make_thunk(node,
+                                                  storage_map,
+                                                  compute_map,
+                                                  no_recycling)
         return ret
 
+    def init_c_code(self):
+        """Return the C code for this OpFromGraph.  """
+        subd = dict(chain(
+            ((e, "%%(i%i)s" % i) for i, e in enumerate(self.fgraph.inputs)),
+            ((e, "%%(o%i)s" % i) for i, e in enumerate(self.fgraph.outputs))))
+
+        for var in self.fgraph.variables:
+            if var.owner is None:
+                if var not in self.fgraph.inputs:
+                    # This is an orphan
+                    if isinstance(var, Constant):
+                        subd[var] = var.type.c_literal(var.data)
+                    else:
+                        raise ValueError(
+                            "All orphans in the fgraph to Composite must"
+                            " be Constant instances.")
+            elif (any(i.dtype == 'float16' for i in var.owner.inputs) or
+                  any(o.dtype == 'float16' for o in var.owner.outputs)):
+                # flag for elemwise ops to check.
+                self.inner_float16 = True
+
+        _c_code = "{\n"
+        self.nodenames = ["%(nodename)s_" + ('subnode%i' % j)
+                          for j, n in enumerate(self.fgraph.toposort())]
+
+        i = 0
+        for j, node in enumerate(self.fgraph.toposort()):
+            for output in node.outputs:
+                if output not in subd:
+                    i += 1
+                    name = "V%%(id)s_tmp%i" % i
+                    subd[output] = name
+                    _c_code += "%s %s;\n" % (
+                            output.type.dtype_specs()[1], name)
+            s = node.op.c_code(
+                node,
+                self.nodenames[j],
+                [subd[input] for input in node.inputs],
+                [subd[output] for output in node.outputs],
+                dict(fail="%(fail)s", id="%%(id)s_%i" % j))
+            _c_code += s
+            _c_code += "\n"
+        _c_code += "}\n"
+
+        for char in 'idflxsc':
+            _c_code = re.sub('\%'+char, "%%"+char, _c_code)
+
+        self._c_code = _c_code
```

Comment:
This should share code with CLinker.  Otherwise it's bound to become obsolete and not work anymore sometime soon.